src/Pos/Wallet/Web/BListener.hs

```diff
@@ -63,10 +63,17 @@ onApplyTracking
     => OldestFirst NE (Blund ssc) -> m SomeBatchOp
 onApplyTracking blunds = do
     let oldestFirst = getOldestFirst blunds
-        txs = concatMap (gbTxs . fst) oldestFirst
-        newTipH = NE.last oldestFirst ^. _1 . blockHeader
-    -- AJ: TODO: Efficiency
-    sd <- HM.fromList <$> GS.getAllSlottingData
+        txs         = concatMap (gbTxs . fst) oldestFirst
+        newTipH     = NE.last oldestFirst ^. _1 . blockHeader
+        newEpoch    = NE.last oldestFirst ^. _1 . epochIndexL
+
+    -- Apply blocks only from last epoch, since this can be called only with blocks
+    -- from a single epoch.
+    sd <- do
```

Comment:
It should be done differently. Instead of fetching whole slotting data from DB and then filtering it, you can iterate over DB and pick only desired data. Please don't do it before rebasing to master, because there everything changed a lot.